api_app/connectors_manager/connectors/misp.py

```diff
@@ -1,10 +1,128 @@
+
+        # Add tags from Job
+        for tag in self._job.tags.all():
+            obj.add_tag(f"intelowl-tag:{tag.label}")
+
+        return obj
+
+    def _get_attr_obj(self, type, value) -> pymisp.MISPAttribute:
+        obj = pymisp.MISPAttribute()
+        obj.type = type
+        obj.value = value
+        return obj
+
+    @property
+    def _base_attr_obj(self) -> pymisp.MISPAttribute:
+        if self._job.is_sample:
+            type = INTELOWL_MISP_TYPE_MAP["file"]
+            binary = helpers.get_binary(self.job_id)
+            md5 = hashlib.md5(binary).hexdigest()
+            value = f"{self._job.file_name}|{md5}"
+        else:
+            type = self._job.observable_classification
+            value = self._job.observable_name
+            if type == "hash":
+                matched_type = helpers.get_hash_type(value)
+                type = matched_type if matched_type is not None else "text"
+            else:
+                type = INTELOWL_MISP_TYPE_MAP[type]
+
+        obj = self._get_attr_obj(type, value)
+        obj.comment = f"Analyzers Executed: {self._job.analyzers_to_execute}"
+        return obj
+
+    @property
+    def _secondary_attr_objs(self) -> List[pymisp.MISPAttribute]:
+        obj_list = []
+        if self._job.is_sample:
+            # mime-type
+            obj_list.append(self._get_attr_obj("mime-type", self._job.file_mimetype))
+        return obj_list
+
+    @property
+    def _link_attr_obj(self) -> pymisp.MISPAttribute:
+        """
+        Returns attribute linking analysis on IntelOwl instance
+        """
+        obj = pymisp.MISPAttribute()
+        obj.type = "link"
+        obj.value = f"{settings.WEB_CLIENT_URL}/pages/scan/result/{self.job_id}"
+        obj.comment = "View Analysis on IntelOwl"
+        obj.disable_correlation = True
+
+        return obj
+
     def run(self):
-        return {"success": True}
+        misp_instance = pymisp.PyMISP(
+            url=self.__url_name,
+            key=self.__api_key,
+            ssl=self.ssl_check,
+            debug=self.debug,
+            timeout=5,
+        )
+
+        # get event and attributes
+        event = self._event_obj
+        attributes = [
+            self._base_attr_obj,
+            *self._secondary_attr_objs,
+            self._link_attr_obj,
+        ]
+
+        # append attribute name to event info
+        event.info += f": {self._base_attr_obj.value}"
+
+        # add event to MISP Instance
+        misp_event = misp_instance.add_event(event, pythonify=True)
+        # add attributes to event on MISP Instance
+        for attr in attributes:
+            misp_instance.add_attribute(misp_event.id, attr)
+
+        return misp_instance.get_event(misp_event.id)
+
+    @classmethod
+    def _monkeypatch(cls):
+        # completely skip because doesnt work without connection.
+        patches = [if_mock_connections(patch.object(cls, "run", return_value={}))]
```

Comment:
to avoid to have 0 coverage for this code, is it possible to try to make a more advanced patching?

you can see as an example how we patched the MWdb library [here](https://github.com/intelowlproject/IntelOwl/blob/develop-2/api_app/analyzers_manager/file_analyzers/mwdb_scan.py#L18)

Is it enough to just set the attributes and methods required by our code. In this way all the logic unrelated to the library pymisp would be correctly tested.

Reply:
done, I've added a mock class only for the methods which connect with the instance.